fe/fe-core/src/main/java/com/starrocks/sql/ShortCircuitPlanner.java

```diff
@@ -0,0 +1,175 @@
+import com.starrocks.sql.optimizer.operator.Operator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalFilterOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalOlapScanOperator;
+import com.starrocks.sql.optimizer.operator.logical.LogicalProjectOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ColumnRefOperator;
+import com.starrocks.sql.optimizer.operator.scalar.ScalarOperator;
+
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+
+public class ShortCircuitPlanner {
+
+    public static final long MAX_RETURN_ROWS = 1024;
+
+    public static BaseLogicalPlanChecker createLogicalPlanChecker(OptExpression root, boolean allowFilter,
+                                                                  boolean allowLimit, boolean allowProject,
+                                                                  boolean allowSort, ScalarOperator predicate,
+                                                                  List<String> orderByColumns, long limit) {
+        if (root.getOp() instanceof LogicalOlapScanOperator) {
+            return new ShortCircuitPlannerHybrid.LogicalPlanChecker(allowFilter, allowLimit, allowProject,
+                    allowSort, predicate, orderByColumns, limit);
+        } else {
+            return new BaseLogicalPlanChecker();
+        }
+    }
+
+    public static OptExpression checkSupportShortCircuitRead(OptExpression root, ConnectContext connectContext) {
+        if (!connectContext.getSessionVariable().isEnableShortCircuit()) {
+            root.setShortCircuit(false);
+        }
+        boolean supportShortCircuit = root.getOp().accept(new LogicalPlanChecker(), root, null);
+        root.setShortCircuit(supportShortCircuit);
+        return root;
+    }
+
+    protected static boolean isRedundant(Map<ColumnRefOperator, ScalarOperator> projections) {
+        for (Map.Entry<ColumnRefOperator, ScalarOperator> entry : projections.entrySet()) {
+            if (!entry.getKey().equals(entry.getValue())) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    public static class BaseLogicalPlanChecker extends OptExpressionVisitor<Boolean, Void> {
+        @Override
+        public Boolean visit(OptExpression optExpression, Void context) {
+            return false;
+        }
+
+        protected boolean visitChild(OptExpression optExpression, Void context) {
+            if (optExpression.getInputs().size() != 1) {
+                return false;
+            }
+            OptExpression child = optExpression.getInputs().get(0);
+            return child.getOp().accept(this, child, context);
+        }
+
+    }
+
+    /**
+     * Checker for select plan:
+     * 1. Pattern: Limit - Sort (TopN) - Project - Filter - Scan.
+     * 2. Scan: single row store table.
+     * 3. Filter: contains all keys in table primary keys.
+     * 4. Limit: no offset.
+     */
```

Comment:
Should we process limitOperator here?

Reply:
currently can't process limit operator for column with row， comment is redundant